model-optimizer/mo/graph/port.py

```diff
@@ -135,24 +135,21 @@ def _set_value(self, value):
         if self.node.graph.stage == 'front':
             raise Error("set_value is not applicable for graph front phase")
         else:
-            if self.type == 'in':
-                data_node = self.node.in_node(self.idx, control_flow=self.control_flow)
-                const_node = data_node.in_node(control_flow=self.control_flow)
+            data_node_caller = self.node.in_node if self.type == 'in' else self.node.out_node
+            data_node = data_node_caller(self.idx, control_flow=self.control_flow)
+            const_node = data_node.in_node(control_flow=self.control_flow) if self.type == 'in' else self.node
 
-                # Set value to data node
-                data_node.value = value
-                data_node.shape = int64_array(value.shape)
+            force_shape = data_node.soft_get('force_shape', const_node.soft_get('force_shape', None))
+            shape = int64_array(value.shape if force_shape is None else force_shape)
 
-                # Set value to constant producer
-                if const_node.soft_get('type') == 'Const':
-                    const_node.value = value
-                    const_node.shape = int64_array(value.shape)
-            else:
-                self.node.out_node(self.idx, control_flow=self.control_flow).value = value
-                self.node.out_node(self.idx, control_flow=self.control_flow).shape = int64_array(value.shape)
-                if self.node.soft_get('type') == 'Const':
-                    self.node.value = value
-                    self.node.shape = int64_array(value.shape)
+            # Set value to data node
+            data_node.value = value
+            data_node.shape = shape
+
+            # Set value to constant producer
+            if const_node.soft_get('type') == 'Const':
+                const_node.value = value
+                const_node.shape = shape
```

Comment:
Please, add several unit tests for a new logic.

Reply:
Done